### PR TITLE
pipeline-aop refactor to provider

### PR DESCRIPTION
**Pull request by ZLValien**
#### What type of this PR

/kind feature

#### What this PR does / why we need it:

refactor aop.

now, the tuneType and tuneTrigger of plugins is  defined in pipeline.yml。

https://erda.cloud/erda/dop/projects/387/issues/task?id=216982&issueFilter__urlQuery=eyJpdGVyYXRpb25JRHMiOls1MDZdLCJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciXSwiYXNzaWduZWVJRHMiOlsiMTAwMDgzMiJdfQ%3D%3D&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoic3RhdHVzIn19&iterationID=506&type=TASK

### CODE REVIEW

File: modules/pipeline/pipengine/reconciler/queuemanage/manager/define.go [hunk 1]
```diff
@@ -31,15 +32,18 @@ type defaultManager struct {
...
 
 	dbClient *dbclient.Client
+
+	// aop
+	pluginsManage *plugins_manage.PluginsManage
 }
 
 // New return a new queue manager.
-func New(ops ...Option) types.QueueManager {
+func New(pluginsManage *plugins_manage.PluginsManage, ops ...Option) types.QueueManager {
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/queuemanage/manager/define.go:41]: use ops at least.

File: modules/pipeline/pipengine/reconciler/queuemanage/queue/define.go [hunk 2]
```diff
@@ -55,15 +55,19 @@ type defaultQueue struct {
 
 	// is updating pending queue
 	updatingPendingQueue bool
+
+	// aop
+	PluginsManage *plugins_manage.PluginsManage
 }
 
-func New(pq *apistructs.PipelineQueue, ops ...Option) *defaultQueue {
+func New(pq *apistructs.PipelineQueue, pm *plugins_manage.PluginsManage, ops ...Option) *defaultQueue {
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/queuemanage/queue/define.go:63]: so ugly.

File: modules/pipeline/providers/aop/plugins/pipeline/apitest_report/plugin.go [hunk 3]
```diff
@@ -153,3 +155,34 @@ func New() *Plugin {
...
+}
+
+// +provider
+type provider struct {
+	Cfg *config
+}
+
+func (p *provider) Init(ctx servicehub.Context) error {
+	for _, tuneTrigger := range p.Cfg.TuneTrigger {
+		err := plugins_manage.RegisterTunePointToTuneGroup(p.Cfg.TuneType, tuneTrigger, New())
```
Comment (sfwn) [inline modules/pipeline/providers/aop/plugins/pipeline/apitest_report/plugin.go:171]: Use `aop.Register` and register to pkg level variables.

File: modules/pipeline/providers/aop/README.md [hunk 4]
```diff
@@ -1,26 +1,30 @@
...
+- 在 `tunechain.go` 对应的触发时机下编排你的插件
 
-import (
-	"github.com/erda-project/erda/modules/pipeline/aop/aoptypes"
-	"github.com/erda-project/erda/modules/pipeline/aop/plugins/task/plugins/autotest_cookie_keep_after"
-	"github.com/erda-project/erda/modules/pipeline/aop/plugins/task/plugins/autotest_cookie_keep_before"
-	"github.com/erda-project/erda/modules/pipeline/aop/plugins/task/plugins/echo"
-	"github.com/erda-project/erda/modules/pipeline/aop/plugins/task/plugins/unit_test_report"
-)
+// TuneTriggerChains 保存流水线所有触发时机下的调用链
```
Comment (sfwn) [inline modules/pipeline/providers/aop/README.md:8]: remove codes if wrong.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: conf/pipeline/pipeline.yaml [hunk 5]
```diff
@@ -18,3 +18,67 @@ erda.core.pipeline.cms:
...
+erda.core.pipeline.aop:
+  chains:
+    pipeline:
+      pipeline_before_exec:
+        - "echo"
```
Comment (sfwn) [inline conf/pipeline/pipeline.yaml:25]: I think now we can remove useless `echo` plugin

File: conf/pipeline/pipeline.yaml [hunk 6]
```diff
@@ -18,3 +18,67 @@ erda.core.pipeline.cms:
...
+      task_before_wait:
+        - "echo"
+      task_after_wait:
+        - "echo"
+
+erda.core.pipeline.aop.plugins.pipeline.echo:
+
+erda.core.pipeline.aop.plugins.pipeline.project:
+
```
Comment (sfwn) [inline conf/pipeline/pipeline.yaml:67]: no need new line.

File: modules/pipeline/aop/context.go [hunk 7]
```diff
@@ -1,51 +0,0 @@
-// Copyright (c) 2021 Terminus, Inc.
```
Comment (sfwn) [inline modules/pipeline/aop/context.go:1]: DO NOT delete these files.

File: modules/pipeline/providers/aop/plugins/pipeline/apitest_report/plugin.go [hunk 8]
```diff
@@ -153,3 +155,33 @@ func New() *Plugin {
 	var p Plugin
 	return &p
 }
+
+type config struct {
+	TuneType    aoptypes.TuneType      `file:"tune_type"`
```
Comment (sfwn) [inline modules/pipeline/providers/aop/plugins/pipeline/apitest_report/plugin.go:160]: useless configs?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: cmd/pipeline/main.go [hunk 9]
```diff
@@ -22,6 +22,8 @@ import (
 	_ "github.com/erda-project/erda-infra/providers/mysqlxorm"
 	_ "github.com/erda-project/erda-infra/providers/serviceregister"
 	_ "github.com/erda-project/erda/modules/pipeline"
+	_ "github.com/erda-project/erda/modules/pipeline/providers/aop"
+	_ "github.com/erda-project/erda/modules/pipeline/providers/aop/plugins/all"
```
[the remark this answers is not in the document]
Comment (ZLValien) [reply cmd/pipeline/main.go]: if all.go add to  pkg `github.com/erda-project/erda/modules/pipeline/providers/aop` will  cycle import.
I add it to `github.com/erda-project/erda/modules/pipeline`

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/aop/aoptypes/tunegroup.go [hunk 10]
```diff
@@ -14,10 +14,13 @@
 
 package aoptypes
 
-// TuneGroup 保存所有类型不同触发时机下的调用链
+// TunePointNameGroup Save the TunePointName of all types of call chains under different trigger
+type TunePointNameGroup map[TuneType]map[TuneTrigger][]string
```
Comment (sfwn) [inline modules/pipeline/aop/aoptypes/tunegroup.go:18]: Move this struct definition to where it used. Here is strange.

File: modules/pipeline/aop/aoptypes/tunegroup.go [hunk 11]
```diff
@@ -14,10 +14,13 @@
...
+
+// TuneGroup Save all types of call chains under different trigger
 type TuneGroup map[TuneType]map[TuneTrigger]TuneChain
 
-// GetTuneChainByTypeAndTrigger 根据 类型 和 触发时机 返回 调用链
+// GetTuneChainByTypeAndTrigger By TuneType and TuneTrigger return TuneChain
```
Comment (sfwn) [inline modules/pipeline/aop/aoptypes/tunegroup.go:23]: `return tuneChain by xxx`

let your words smooth.

File: modules/pipeline/aop/plugins/pipeline/precheck_before_pop/plugin.go [hunk 12]
```diff
@@ -25,7 +27,7 @@ type Plugin struct {
 
 func New() *Plugin { return &Plugin{} }
```
Comment (sfwn) [inline modules/pipeline/aop/plugins/pipeline/precheck_before_pop/plugin.go:28]: use provider struct to implements this interface.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/aop/register.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+func NewProviderNameByPluginName(point TunePoint) string {
+	// please use "-" replace all "_"
+	if strings.Contains(point.Name(), "_") {
+		panic("provider name invaild, tuneType: " + string(point.Type()) + ", name: " + point.Name())
```
Comment (sfwn) [inline modules/pipeline/aop/register.go:72]: tell the reason

File: modules/pipeline/aop/plugins/pipeline/apitest_report/plugin.go [hunk 14]
```diff
@@ -29,22 +31,18 @@ const (
...
-
 type ApiReportMeta struct {
 	ApiTotalNum   int `json:"apiTotalNum"`
 	ApiSuccessNum int `json:"apiSuccessNum"`
 	ApiFailedNum  int `json:"apiFailedNum"`
 	ApiNotExecNum int `json:"apiNotExecNum"`
 }
 
-func (p *Plugin) Handle(ctx *aoptypes.TuneContext) error {
+func New() *provider { return &provider{} }
```
Comment (sfwn) [inline modules/pipeline/aop/plugins/pipeline/apitest_report/plugin.go:41]: No need exported func `New()`.
`aop.NewProviderNameByPluginName(&provider{})` is enough.

File: modules/pipeline/aop/plugins/pipeline/apitest_report/plugin.go [hunk 15]
```diff
@@ -149,7 +147,23 @@ func (p *Plugin) Handle(ctx *aoptypes.TuneContext) error {
 	return nil
 }
 
-func New() *Plugin {
-	var p Plugin
-	return &p
+// +provider
+type provider struct {
```
Comment (sfwn) [inline modules/pipeline/aop/plugins/pipeline/apitest_report/plugin.go:151]: move struct define to beginning, the same as `Plugin` before.

File: modules/pipeline/aop/plugins/pipeline/apitest_report/plugin.go [hunk 16]
```diff
@@ -149,7 +147,23 @@ func (p *Plugin) Handle(ctx *aoptypes.TuneContext) error {
...
+	aoptypes.PipelineBaseTunePoint
+}
+
+func (p *provider) Init(ctx servicehub.Context) error {
+	err := aop.RegisterTunePoint(p)
```
Comment (sfwn) [inline modules/pipeline/aop/plugins/pipeline/apitest_report/plugin.go:156]: if err := xxx; err != nil

File: modules/pipeline/aop/register.go [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+	if !ok {
+		group = make(map[TuneTrigger]TuneChain)
+	}
+	group[tuneTrigger] = append(group[tuneTrigger], tunePoint)
+	tuneGroup[tuneType] = group
+	return nil
+}
+
+func ConvertTuneGroup(gp TunePointNameGroup) error {
+	// Convert TunePointNameGroup to TuneGroup
```
Comment (sfwn) [inline modules/pipeline/aop/register.go:55]: Move to func comment.

File: modules/pipeline/aop/register.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+// RegisterTunePoint register tunePoint
+func RegisterTunePoint(tunePoint TunePoint) error {
+	if _, ok := pluginsMap[tunePoint.Type()]; !ok {
+		pluginsMap[tunePoint.Type()] = make(map[string]TunePoint)
+	}
+
+	// if tunePoint name is existed
+	if _, ok := pluginsMap[tunePoint.Type()][tunePoint.Name()]; ok {
+		return errors.New("tunePoint is existed, tuneType: " + string(tunePoint.Type()) + ", Name: " + tunePoint.Name())
```
Comment (sfwn) [inline modules/pipeline/aop/register.go:34]: failed to register xxx, tunePoint is already exist, xxx

File: modules/pipeline/aop/register.go [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+	group, ok := tuneGroup[tuneType]
+	if !ok {
+		group = make(map[TuneTrigger]TuneChain)
+	}
+	group[tuneTrigger] = append(group[tuneTrigger], tunePoint)
+	tuneGroup[tuneType] = group
+	return nil
+}
+
+func ConvertTuneGroup(gp TunePointNameGroup) error {
```
Comment (sfwn) [inline modules/pipeline/aop/register.go:54]: no need export this func.

Comment (sfwn) [reply modules/pipeline/aop/register.go:54]: bad name.
`handleTuneConfigs`

File: modules/pipeline/aop/provider.go [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	"context"
+
+	"github.com/erda-project/erda-infra/base/logs"
+	"github.com/erda-project/erda-infra/base/servicehub"
+	"github.com/erda-project/erda/modules/pipeline/aop/aoptypes"
+)
+
+// TunePointNameGroup Save the TunePointName of all types of call chains under different trigger
+type TunePointNameGroup map[aoptypes.TuneType]map[aoptypes.TuneTrigger][]string
```
Comment (sfwn) [inline modules/pipeline/aop/provider.go:26]: TuneConfigs

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**